### PR TITLE
Refactor zig fmt indentation. Remove indent from rendering code

### PR DESCRIPTION
**Pull request by LakeByTheWoods**
Creating new pull as I couldn't work out how to open my ancient one.
I have been on and off rebasing and updating this code throughout the year and believe it is the right way to go.
The change appears larger than it is due to the removal of `indent, start_col` from so many call sights. There is bound to be merge conflict with people's local work on render.zig, I am happy to work through these conflicts as they come.

This (I hope) will make dealing with indentation in zig fmt easier to work with. See #4580 (TLDR: remove indent and start_col, replace with an auto-indenting-stream)
The auto indenting stream is more complicated than the original proposal to cover the edge cases of all the places zig wants to put indentation.

### CODE REVIEW

File: lib/std/io/auto_indenting_stream.zig [hunk 1]
```diff
@@ -0,0 +1,135 @@
+const std = @import("../std.zig");
+const io = std.io;
+const mem = std.mem;
+const assert = std.debug.assert;
+
+pub fn AutoIndentingStream(comptime indent_delta: u8, comptime OutStreamType: type) type {
+    return struct {
+        const Self = @This();
+        pub const Error = OutStreamType.Error;
+        pub const OutStream = io.Writer(*Self, Error, write);
```
Comment (daurnimator) [inline lib/std/io/auto_indenting_stream.zig:10]: Call this `Writer`

File: lib/std/io/auto_indenting_stream.zig [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+        /// Push default indentation
+        pub fn pushIndent(self: *Self) void {
+            // Doesn't actually write any indentation. Just primes the stream to be able to write the correct indentation if it needs to.
+            self.pushIndentN(indent_delta);
+        }
+
+        /// Push an indent of arbitrary width
+        pub fn pushIndentN(self: *Self, n: u8) void {
+            assert(self.indent_stack_top < std.math.maxInt(u8));
```
Comment (daurnimator) [inline lib/std/io/auto_indenting_stream.zig:66]: so `zig fmt` will panic/exhibit undefined behaviour if you have > 255 levels of indentation?

File: lib/std/io/auto_indenting_stream.zig [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        /// Push default indentation
+        pub fn pushIndent(self: *Self) void {
+            // Doesn't actually write any indentation. Just primes the stream to be able to write the correct indentation if it needs to.
+            self.pushIndentN(indent_delta);
+        }
+
+        /// Push an indent of arbitrary width
+        pub fn pushIndentN(self: *Self, n: u8) void {
```
Comment (daurnimator) [inline lib/std/io/auto_indenting_stream.zig:65]: I find myself asking "why `u8`"?

File: lib/std/io/change_detection_stream.zig [hunk 4]
```diff
@@ -0,0 +1,58 @@
+const std = @import("../std.zig");
+const io = std.io;
+const mem = std.mem;
+const assert = std.debug.assert;
+
+pub fn ChangeDetectionStream(comptime OutStreamType: type) type {
```
Comment (daurnimator) [inline lib/std/io/change_detection_stream.zig:6]: Could you add a doccomment of what this does?

File: lib/std/io/change_detection_stream.zig [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+        pub fn init(source: []const u8, out_stream: *OutStreamType) Self {
+            return Self{
+                .out_stream = out_stream,
+                .source_index = 0,
+                .source = source,
+            };
+        }
+
+        pub fn outStream(self: *Self) OutStream {
```
Comment (daurnimator) [inline lib/std/io/change_detection_stream.zig:25]: should be `writer`

File: lib/std/io/find_byte_out_stream.zig [hunk 6]
```diff
@@ -0,0 +1,44 @@
+const std = @import("../std.zig");
+const io = std.io;
+const assert = std.debug.assert;
+
+// An OutStream that returns whether the given character has been written to it.
```
Comment (daurnimator) [inline lib/std/io/find_byte_out_stream.zig:5]: should be doccomment

**Review by daurnimator — CHANGES_REQUESTED** [review]

File: lib/std/io/auto_indenting_stream.zig (hunk 2 above)
Comment (LakeByTheWoods) [reply lib/std/io/auto_indenting_stream.zig:66]: Yeah, I guess that could possibly maybe happen with generated code. Worth bumping to a u32?

File: lib/std/io/auto_indenting_stream.zig (hunk 3 above)
Comment (LakeByTheWoods) [reply lib/std/io/auto_indenting_stream.zig:65]: I don't have an answer I can give with conviction

File: lib/std/io/change_detection_stream.zig (hunk 4 above)
Comment (LakeByTheWoods) [reply lib/std/io/change_detection_stream.zig:6]: Sure can

File: lib/std/io/auto_indenting_stream.zig (hunk 2 above)
Comment (LakeByTheWoods) [reply lib/std/io/auto_indenting_stream.zig:66]: Though looking again, if we want to store a large number of indentations we would then need to dynamically allocate and pass in an allocator.

File: lib/std/io/auto_indenting_stream.zig [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            if (bytes.len == 0)
+                return @as(usize, 0);
+
+            try self.applyIndent();
+            return self.writeNoIndent(bytes);
+        }
+
+        fn writeNoIndent(self: *Self, bytes: []const u8) Error!usize {
+            try self.out_stream.outStream().writeAll(bytes);
+            if (bytes[bytes.len - 1] == '\n')
```
Comment (LemonBoy) [inline lib/std/io/auto_indenting_stream.zig:38]: Missing guard for `bytes.len == 0`

Comment (LakeByTheWoods) [reply lib/std/io/auto_indenting_stream.zig:38]: Good catch

**Comment by LakeByTheWoods** [comment]
Note: I have more fixes for several issues pending that rely on this review  

https://github.com/LakeByTheWoods/zig/tree/fmt_fixes

**Review by daurnimator — APPROVED** [review]

File: lib/std/io/auto_indenting_stream.zig (hunk 2 above)
Comment (data-man) [reply lib/std/io/auto_indenting_stream.zig]: Can be used LinearFifo here? With `{ .Static }` option.

File: lib/std/zig/parser_test.zig [hunk 8]
```diff
@@ -2885,20 +2964,20 @@ test "zig fmt: multiline string in array" {
...
-        \\,
+        \\    ,
         \\    \\bbb
         \\};
         \\
         \\fn bar() void {
         \\    const Foo = [][]const u8{
         \\        \\aaa
-        \\    ,
+        \\        ,
```
Comment (andrewrk) [inline lib/std/zig/parser_test.zig:2974]: This looks inconsistent with the newly added test `test "zig fmt: Only indent multiline string literals in function calls"`

File: lib/std/io.zig [hunk 9]
```diff
@@ -182,10 +191,10 @@ pub const BufferedAtomicFile = @import("io/buffered_atomic_file.zig").BufferedAt
 pub const StreamSource = @import("io/stream_source.zig").StreamSource;
 
 /// A Writer that doesn't write to anything.
-pub const null_writer = @as(NullWriter, .{ .context = {} });
+pub var null_writer = @as(NullWriter, .{ .context = {} });
```
Comment (andrewrk) [inline lib/std/io.zig:194]: Can you talk a little bit about the idea here to change streams from being passed by-val? Why is this necessary?

File: lib/std/io/auto_indenting_stream.zig [hunk 10]
```diff
@@ -0,0 +1,150 @@
+const std = @import("../std.zig");
+const io = std.io;
+const mem = std.mem;
+const Allocator = mem.Allocator;
+const ArrayList = std.ArrayList;
+const assert = std.debug.assert;
+
+/// Automatically inserts indentation of written data by keeping
+/// track of the current indentation level
+pub fn AutoIndentingStream(comptime indent_delta: usize, comptime WriterType: type) type {
```
Comment (andrewrk) [inline lib/std/io/auto_indenting_stream.zig:10]: Any reason for `indent_delta` to be comptime known? We can reduce code bloat by making this initialized at runtime if possible.

File: lib/std/io/auto_indenting_stream.zig [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    return struct {
+        const Self = @This();
+        pub const Error = WriterType.Error;
+        pub const PushError = Allocator.Error;
+        pub const Writer = io.Writer(*Self, Error, write);
+        const Stack = ArrayList(usize);
+
+        writer_pointer: *WriterType,
+        indent_stack: Stack,
```
Comment (andrewrk) [inline lib/std/io/auto_indenting_stream.zig:19]: This abstraction introduces heap allocation where it really shouldn't be necessary. This is a regression in terms of resource utilization, and likely also performance. I think you can keep this same abstraction, but if you take away the need to support variable indentation widths, you can remove the need for heap allocation like this. As far as I'm aware, 100% of the API users of this abstraction don't need variable indentation widths.

Comment (LakeByTheWoods) [reply lib/std/io/auto_indenting_stream.zig:19]: I *almost* agree 100%, there's just the one place that needs a different indentation and that's the two space indentation in the assembly block. I'll have a think of a way of supporting it without the need of tracking the stack.

File: lib/std/io/auto_indenting_stream.zig (hunk 10 above)
Comment (LakeByTheWoods) [reply lib/std/io/auto_indenting_stream.zig:10]: No special reason, it was just a comptime value prior to the indenting stream. Will change.

File: lib/std/io.zig (hunk 9 above)
Comment (LakeByTheWoods) [reply lib/std/io.zig:194]: Oh, that's a left over as it turns out, I'll revert it. As the auto_indenting_stream needs to be non-const there were some awkward const dependencies going on.

File: lib/std/zig/parser_test.zig (hunk 8 above)
Comment (LakeByTheWoods) [reply lib/std/zig/parser_test.zig:2974]: Yeah, this is an awkward situation. Everything gets indented inside the `{}` in the above, but in the "Only indent" test nothing gets indented. Because there's no trailing space it's kind-of a pseudo-single-line function call, so a special case is made for multiline strings. If #5343 they could trivially be made to both indent the comma by simply indenting everything in the function call (NB: this would shift in the comma in all testTransforms). To unindent the above comma, as there would be a special case for the comma after a multiline string in an array init.

File: lib/std/io/auto_indenting_stream.zig (hunk 11 above)
Comment (andrewrk) [reply lib/std/io/auto_indenting_stream.zig:19]: > two space indentation in the assembly block

We can change the rules on this if it means that `zig fmt` becomes fundamentally simpler to implement. Feel free to do that.

Comment (LakeByTheWoods) [reply lib/std/io/auto_indenting_stream.zig:19]: Done some basic performance testing, running fmt on std takes 3.5secs before the change and 4.2secs after. With a few modifications I found that all of that lost time can be gotten back by replacing the indent stack with a counter. So definitely worth fixing.

File: lib/std/zig/parser_test.zig (hunk 8 above)
Comment (andrewrk) [reply lib/std/zig/parser_test.zig:2974]: At this point, you've invested enough time and thought on the subject that I would trust your judgement. So this will be your call.

File: lib/std/io/auto_indenting_stream.zig (hunk 11 above)
Comment (LakeByTheWoods) [reply lib/std/io/auto_indenting_stream.zig]: I managed to keep to 2 space indent setting the indentation level to 2 when rendering the inline asm elements. Not the *most* elegant solution but it works without compromising the current style.

File: src-self-hosted/main.zig [hunk 12]
```diff
@@ -682,13 +682,13 @@ pub fn cmdFmt(gpa: *Allocator, args: []const []const u8) !void {
             process.exit(1);
         }
         if (check_flag) {
-            const anything_changed = try std.zig.render(gpa, io.null_out_stream, tree);
+            const anything_changed = try std.zig.render(gpa, &io.null_out_stream, tree);
```
Comment (andrewrk) [inline src-self-hosted/main.zig:685]: The `&` on these can be reverted now right? Other than that this looks ready to merge now.

**Comment by LakeByTheWoods** [comment]
I've done some rudimentary perf testing, there's definitely been a drop.
Typical before:

```
perf stat -e cache-misses,instructions,cycles,branch-misses zig fmt .

 Performance counter stats for 'zig fmt .':

        36,326,992      cache-misses:u                                              
    95,117,916,592      instructions:u            #    2.89  insn per cycle         
    32,936,634,845      cycles:u                                                    
        90,260,291      branch-misses:u                                             

      10.332502466 seconds time elapsed

       9.837209000 seconds user
       0.469674000 seconds sys
```



Typical after:
```perf stat -e cache-misses,instructions,cycles,branch-misses zig fmt .

 Performance counter stats for 'zig fmt .':

        36,023,687      cache-misses:u                                              
   115,786,315,719      instructions:u            #    3.11  insn per cycle         
    37,241,633,937      cycles:u                                                    
        95,507,538      branch-misses:u                                             

      11.621999069 seconds time elapsed

      11.094600000 seconds user
       0.499317000 seconds sys
```

data was a folder with 27 copies of lib/zig/std

**Comment by LakeByTheWoods** [comment]
Oh, and thanks for the patch!

**Comment by andrewrk** [comment]
> 27 copies of lib/zig/std

:open_mouth: 

Thanks for the perf report! It really helps to see this kind of thing with a big change like this, even if it does show a small regression. -12% is probably right on the tolerance level for this, but I think it's OK to make this kind of tradeoff when we still have a few bugs and issues to resolve, making investments into cleaning up the code structure and abstractions worthwhile.

**Comment by marnix** [comment]
@LakeByTheWoods Not sure if intentional, but for your information: https://github.com/marnix/zigmmverify/pull/19/commits/7a4aec9be8d8a9e8b62717d1bfdc54376125399a shows an indent change (improvement, IMO) which was likely caused by this MR.

**Comment by LakeByTheWoods** [comment]
@marnix Well I'll certainly take credit for it :grin: Wasn't so much intentional as it is a happy side-effect.
It's just as likely for there to be sad side-effects from patterns not covered by tests.
